File: 153.find-minimum-in-rotated-sorted-array.java
/*
 * @lc app=leetcode id=153 lang=java
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
class Solution {
    public int findMin(int[] nums) {
        if(nums==null||nums.length==0){
            return 0;
        }
        if(nums.length==1){
            return nums[0];
        }


        int left=0,right=nums.length-1;
        while(left<right){
            int mid=left+(right-left)/2;
            if(mid>0&&nums[mid]<nums[mid-1]){
                return nums[mid];
            }
            if(nums[left]<=nums[mid]&&nums[mid]>nums[right]){
                left=mid+1;
            }else{
                right=mid-1;
            }
        }
        return nums[left];
    }
}
// @lc code=end

File: lookup/153-find-minimum-in-rotated-sorted-array.go
package main

func findMin(nums []int) int {
	left, right := 0, len(nums)-1
	for left < right {
		mid := left + (right-left)>>1
		if nums[mid] < nums[right] {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return nums[left]
}

//func main() {
//	nums := []int{3, 4, 5, 1, 2}
//	fmt.Printf("%v \n", findMin(nums))
//}
